### PR TITLE
Don't suggest break through nested items

### LINKED ISSUE #112020: break loop help ignores nested closures
### Code

```Rust
fn test() {
    loop {
        let closure = || {
            if true {
                Err(1)
            }
            
            Ok(())
        };
    }
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:5:17
  |
4 | /             if true {
5 | |                 Err(1)
  | |                 ^^^^^^ expected `()`, found `Result<_, {integer}>`
6 | |             }
  | |_____________- expected this to be `()`
  |
  = note: expected unit type `()`
                  found enum `Result<_, {integer}>`
help: you might have meant to break the loop with this value
  |
5 |                 break Err(1);
  |                 +++++       +
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:4:13
  |
4 | /         if true {
5 | |             Err(1)
  | |             ^^^^^^ expected `()`, found `Result<_, {integer}>`
6 | |         }
  | |_________- expected this to be `()`
  |
  = note: expected unit type `()`
                  found enum `Result<_, {integer}>`
```

### Rationale and extra context

_No response_

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #112020

### CODE REVIEW

**Comment by SparkyPotato** [comment]
If I'm not mistaken, this still suggests breaking out of the loop inside a closure - not sure what `hir::Node` variant closures have.

In the nested item case, currently, the compiler doesn't suggest breaking out of the loop, but instead returning (but only if the type and return types match):
```rs
error[E0308]: mismatched types
 --> src/lib.rs:5:17
  |
4 | /             if true {
5 | |                 Err(1)
  | |                 ^^^^^^ expected `()`, found `Result<_, {integer}>`
6 | |             }
  | |_____________- expected this to be `()`
  |
  = note: expected unit type `()`
                  found enum `Result<_, {integer}>`
help: you might have meant to return this value
  |
5 |                 return Err(1);
  |                 ++++++       +
```

**Comment by compiler-errors** [comment]
lol, @SparkyPotato I should've read my own UI test street. I had made this work for closures, then refactored it for items and didn't check that I broke it for closures 😭

@rustbot author

**Comment by compiler-errors** [comment]
lol, @SparkyPotato I should've read my own UI test street. I had made this work for closures, then refactored it for items and didn't check that I broke it for closures 😭

@rustbot author

**Comment by compiler-errors** [comment]
> In the nested item case, currently, the compiler doesn't suggest breaking out of the loop

This does suggest breaking out of the loop currently:

```rust
fn b() {
    let _ = loop {
        fn foo() -> Result<(), ()> {
            if true {
                Err(1)
                //~^ ERROR mismatched types
            }
            Err(())
        }
    };
}
```

```
error[[E0308]](https://doc.rust-lang.org/nightly/error_codes/E0308.html): mismatched types
 --> src/lib.rs:5:17
  |
4 | /             if true {
5 | |                 Err(1)
  | |                 ^^^^^^ expected `()`, found `Result<_, {integer}>`
6 | |                 //~^ ERROR mismatched types
7 | |             }
  | |_____________- expected this to be `()`
  |
  = note: expected unit type `()`
                  found enum `Result<_, {integer}>`
help: you might have meant to break the loop with this value
  |
5 |                 break Err(1);
  |                 +++++       +
```

But that's fixed by this PR. 

> In the nested item case, currently, the compiler doesn't suggest breaking out of the loop, but instead returning (but only if the type and return types match)

Can you give a more complete example? Not sure if you're asking for something additional that needs fixing, or just making an observation.

**Comment by SparkyPotato** [comment]
On stable:
```rs
fn test() {
    loop {
        fn x() -> Result<(), i32> {
            if true {
                Err(1)
            }
            
            Ok(())
        }
    }
}
```
suggests returning the value, while:
```rs
fn test() {
    loop {
        fn x() -> Result<(), ()> {
            if true {
                Err(1)
            }
            
            Ok(())
        }
    }
}
```
has no suggestions.

**Comment by compiler-errors** [comment]
> Seemingly, the let _ =  changes the behavior here, while it didn't make a difference for the closure case.

That's because the diagnostic checks that there's a `let` statement in some parent HIR node. I don't think that's necessarily the best heuristic, but if you'd like, you should dive deeper into other ways to make this diagnostic give wonky suggestions and possibly fix those too 😸

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -874,7 +874,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-            || self.tcx.hir().parent_iter(id).any(|(parent_id, _)| self.is_loop(parent_id));
+            || self
+                .tcx
+                .hir()
+                .parent_iter(id)
+                .take_while(|(_, node)| {
+                    !matches!(
+                        node,
+                        hir::Node::Item(_)
+                            | hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Closure(_), .. })
```
Comment (SparkyPotato) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:885]: Just noticed that we should probably also check for `async` blocks.

File: tests/ui/loops/dont-suggest-break-thru-item.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+fn a() {
+    loop {
+        let closure = || {
+            if true {
+                Err(1)
+                //~^ ERROR mismatched types
```
Comment (WaffleLapkin) [inline tests/ui/loops/dont-suggest-break-thru-item.rs:6]: Hm, I would expect this to suggest `return`. (still better than the status quo ig)

**Review by WaffleLapkin — APPROVED** [review]
r=me with the async blocks handled (and probably also const blocks too?)

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:885]: We lower async to closures before we get to this point.

**Comment by compiler-errors** [comment]
Async blocks were already handled, but good catch on const blocks.

@bors r=WaffleLapkin

**Pull request merged**